### ISSUE TITLE
Web Version of Qt Console

### ISSUE THREAD

**Issue by afshin** [jupyter/jupyter-js-notebook#170]
In addition to the notebook interface, we need to implement a web-based version of the console interface. Here are some of the tasks that will get us there:
- [x] Create a `ConsolePanel` as a container that holds a `ConsoleWidget`, similar to the way the notebook works.
- [x] Create a `ConsoleModel` that is similar to a `NotebookModel` but does not have a notion of arbitrary cell insertion.
- [x] The first child of the `ConsoleWidget` needs to be a `RawCellWidget` to hold a banner and the second needs to be a `CodeCellWidget` to hold the prompt.
- [ ] The `ConsoleWidget` needs a set of keyboard bindings to emulate the Qt Console application.
- [x] The execution of the code cell needs to be separated from the notebook because it is common functionality.
- [x] After every `prompt` execution, a new `CodeCell` needs to be generated to become the new `prompt`.
- [ ] The UI needs to be styled to look like the Qt Console application.
- [x] The history functionality of the Qt Console needs to be implemented (i.e. navigating through past commands via arrow keys).
- [x] The contextual tooltip feature of the Qt Console needs to be implemented.
- [ ] The `pager` functionality of leaving the prompt and residing within a different interactive context (_i.e._, entering the interactive help) needs to be implemented.
- [x] The completion functionality (_i.e._, tab completion) of the Qt Console needs to be implemented. This should be done in a way that allows the notebook, or other widgets potentially, to use it as well.

cc @blink1073 @jasongrout

**Comment by ellisonbg** [comment]
Yep, and a lot of this refactoring will also help us in hooking up kernels
to standalone text editors.

On Mon, Apr 11, 2016 at 1:17 PM, A. Darian notifications@github.com wrote:

> In addition to the notebook interface, we need to implement a web-based
> version of the console interface. Here are some of the tasks that will get
> us there:
> - Create a ConsolePanel as a container that holds a ConsoleWidget,
>   similar to the way the notebook works.
> - The first child of the ConsoleWidget needs to be a CodeCell
> - The ConsoleWidget needs a set of keyboard bindings to emulate the Qt
>   Console application.
> - The execution of the code cell needs to be separated from the
>   notebook because it is common functionality.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> https://github.com/jupyter/jupyter-js-notebook/issues/170

## 

Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by blink1073** [comment]
@afshin, 

Up/down at the edge of a codemirror: https://github.com/jupyter/notebook/blob/master/notebook/static/notebook/js/cell.js#L234

**Comment by blink1073** [comment]
It looks like the current notebook is still using the deprecated `payload` shell response for the pager:

https://github.com/jupyter/notebook/blob/master/notebook/static/notebook/js/codecell.js#L347

http://jupyter-client.readthedocs.org/en/latest/messaging.html#execution-results

cc @minrk

**Comment by blink1073** [comment]
@afshin, 

Completion: https://github.com/jupyter/notebook/blob/41d6da235cbf3bcf6d7f818e11a066e0fd12ff8b/notebook/static/notebook/js/completer.js#L173
Tooltip for inspection: https://github.com/jupyter/notebook/blob/41d6da235cbf3bcf6d7f818e11a066e0fd12ff8b/notebook/static/notebook/js/tooltip.js#L189

**Comment by blink1073** [comment]
@minrk, ah, I see now "Payloads are considered deprecated, though their replacement is not yet implemented."

This statement is confusing, I would say it isn't deprecated until there is actually a preferred replacement.

**Comment by blink1073** [comment]
@afshin, it looks like all of these will be relevant to you: http://jupyter-client.readthedocs.org/en/latest/messaging.html#payloads-deprecated

**Comment by afshin** [comment]
@blink1073 you're a monster. Thank you!

**Comment by minrk** [comment]
@blink1073 payloads are something that we don't like, and know that we will remove/replace when we get the chance. Since we know that, we want to avoid things like proposing new payloads, etc., but they still exist because we haven't come up with a replacement for the few things they are used for. If you have a better way to communicate "avoid using these if you can, but that's not always possible," A PR to the doc would be welcome.

**Comment by afshin** [comment]
@minrk in order to cover functionality that the Qt Console currently offers, _e.g._, using the interactive `help` system, we'll need to use payloads. Do you have any input on how we should approach this?

cc: @blink1073 @sccolbert 

![help](https://cloud.githubusercontent.com/assets/159529/14788973/47bf5f36-0b02-11e6-8d0b-66a5534a7ded.gif)

**Comment by jasongrout** [comment]
Does the qtconsole use payloads to implement the interactive help? It looks to me like `help()` just uses `raw_input()` to do help. You can see that if you do `help()` in the Jupyter notebook. It looks like you just have to implement the stdin connection: http://jupyter-client.readthedocs.org/en/latest/messaging.html#messages-on-the-stdin-router-dealer-sockets

**Comment by minrk** [comment]
@jasongrout pager output is a payload. @afshin yes, I think the frontend should support payloads. Sorry for the confusion. The comment in the docs mainly pertains to kernels, where we would like to avoid doing _new_ things with payloads until we can come up with their replacement, express a preference to do things without payloads where there's an option.

**Comment by afshin** [comment]
@minrk Thanks for the clarification!

**Comment by jasongrout** [comment]
Ah, thanks for the clarification, @minrk.

**Comment by afshin** [comment]
FYI, this is the current state of affairs (which now includes history):

![console](https://cloud.githubusercontent.com/assets/159529/14924768/fae8a112-0e3b-11e6-8ec3-f84c4f5f3366.gif)

**Comment by afshin** [comment]
![tooltip](https://cloud.githubusercontent.com/assets/159529/15297813/feeb39ea-1b93-11e6-938e-5ad71978e0a0.gif)

**Comment by Carreau** [comment]
Nitpick from screenshots, `Out[` don't have space before the square bracket to align.

**Comment by afshin** [comment]
@Carreau, thanks for the input. @blink1073 just created this issue for it: https://github.com/jupyter/jupyter-js-notebook/issues/221

**Comment by blink1073** [comment]
The output input, as it were.

**Comment by afshin** [comment]
Current state of affairs:

![completion](https://cloud.githubusercontent.com/assets/159529/15621958/a44d0078-245c-11e6-8172-045a226aeaae.gif)

**Comment by jan-matthis** [comment]
Not sure if this is the right place to throw this in: It would be very handy to have the web console closely integrated with the notebook, e.g. through magic commands. Something like `%qtweb` could spawn a console connected to the current kernel in a new tab or in an overlay. `%qtdebug` might spawn a web console with `%debug` executed inside, so that ipdb can be used with autocompletion and history.

**Comment by blink1073** [comment]
Hi @jan-matthis, thank you, those are both excellent suggestions.  We've moved development over to https://github.com/jupyter/jupyterlab.  @afshin, do you want to migrate the top level comment over there and link to this issue?

**Comment by afshin** [comment]
@blink1073 @jan-matthis Sure, will do. Great suggestions.

**Comment by blink1073** [comment]
Migrated to https://github.com/jupyter/jupyterlab/issues/76
